server/src/main/java/io/druid/server/coordinator/DruidCoordinator.java

```diff
@@ -362,22 +362,30 @@ public String getCurrentLeader()
   }
 
   public void moveSegment(
-      final ImmutableDruidServer fromServer,
-      final ImmutableDruidServer toServer,
-      final String segmentName,
+      ImmutableDruidServer fromServer,
+      ImmutableDruidServer toServer,
+      DataSegment segment,
       final LoadPeonCallback callback
   )
   {
+    if (segment == null) {
+      log.makeAlert(new IAE("Can not move null DataSegment"), "Exception moving null segment").emit();
+      if (callback != null) {
+        callback.execute();
+      }
+    }
+    String segmentName = segment.getIdentifier();
     try {
       if (fromServer.getMetadata().equals(toServer.getMetadata())) {
         throw new IAE("Cannot move [%s] to and from the same server [%s]", segmentName, fromServer.getName());
       }
 
-      final DataSegment segment = fromServer.getSegment(segmentName);
-      if (segment == null) {
-        throw new IAE("Unable to find segment [%s] on server [%s]", segmentName, fromServer.getName());
+      DruidDataSource dataSource = metadataSegmentManager.getInventoryValue(segment.getDataSource());
+      if (dataSource == null) {
+        throw new IAE("Unable to find dataSource for segment [%s] in metadata", segmentName);
       }
 
+      final DataSegment segmentToLoad = dataSource.getSegment(segment.getIdentifier());
```

Comment:
also, can you add some comment as to why DataSegment is re-obtained even if we got it in the method arg?